forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,170 @@
+                    } else {
+                        tgtSA = tgtSA.copyWithNoManaCost();
+
+                        // 118.8c
+                        boolean optional = false;
+                        for (CostPart cost : tgtSA.getPayCosts().getCostParts()) {
+                            if ((cost instanceof CostDiscard || cost instanceof CostReveal)
+                                    && !cost.getType().equals("Card") && !cost.getType().equals("Random")) {
+                                optional = true;
+                                break;
+                            }
+                        }
+                        if (!optional) {
+                            tgtSA.getPayCosts().setMandatory(true);
+                        }
+
+                        if (tgtSA.usesTargeting() && !optional) {
+                            tgtSA.getTargetRestrictions().setMandatory(true);
+                        }
+
+                        tgtSA.setSVar("IsCastFromPlayEffect", "True");
+
+                        if (p.getController().playSaFromPlayEffect(tgtSA)) {
+                            final Card played = tgtSA.getHostCard();
+                            // add remember successfully played here if ever needed
+                            final Zone zone = game.getCardState(played).getZone();
+                            if (!origin.equals(zone)) {
+                                CardZoneTable trigList = new CardZoneTable();
+                                trigList.put(origin.getZoneType(), zone.getZoneType(), game.getCardState(found));
+                                trigList.triggerChangesZoneAll(game, sa);
+                            }
+                        }
+                    }
+                }
+            }
+            if (!play || cancel) {
+                changeZone(new CardCollection(found), ZoneType.Hand, game, sa);
+            }
+        }
+
+        // Put the rest on the bottom in a random order.
+        changeZone(rest, ZoneType.Library, game, sa);
+
+    }
+
+    private void changeZone(CardCollection cards, ZoneType zone, Game game, SpellAbility sa) {
+        CardZoneTable table = new CardZoneTable();
+        Map<AbilityKey, Object> moveParams = AbilityKey.newMap();
+        moveParams.put(AbilityKey.LastStateBattlefield, game.copyLastStateBattlefield());
+        moveParams.put(AbilityKey.LastStateGraveyard, game.copyLastStateGraveyard());
+        int pos = 0;
+
+        if (ZoneType.Library.equals(zone)) { // bottom of library in a random order
+            pos = -1;
+            CardLists.shuffle(cards);
+        }
+
+        for (Card c : cards) {
+            final ZoneType origin = c.getZone().getZoneType();
+
+            Card m = game.getAction().moveTo(zone, c, pos, sa, moveParams);
+
+            if (m != null && !origin.equals(m.getZone().getZoneType())) {
+                table.put(origin, m.getZone().getZoneType(), m);
+            }
+        }
+        table.triggerChangesZoneAll(game, sa);
```

Comment:
only the exiling should be sequential
the move rest is not

Reply:
I think I have this sorted –
I also thought about making line 60 `changeZone(new CardCollection(c), ZoneType.Exile, game, sa)` and removing the CardCollection exiled... but it seems like that would be a bit less computationally efficient?
![image](https://github.com/Card-Forge/forge/assets/103371817/bcddc7de-5d9d-4864-80b8-9659a6245cce)